### PR TITLE
Issue#376 create translations when crowdin notifies us

### PR DESCRIPTION
**Pull request by kstigen**

### CODE REVIEW

File: src/main/scala/io/digitallibrary/bookapi/BookApiProperties.scala [hunk 1]
```diff
@@ -88,6 +88,7 @@ object BookApiProperties extends LazyLogging {
   val CorrelationIdHeader = "X-Correlation-ID"
 
   lazy val CrowdinProjects: Seq[CrowdinProject] = readCrowdinProjects()
+  val CrowidinTranslatorPlaceHolder = "INSERT YOUR NAME HERE"
```
Comment (runarfu) [inline src/main/scala/io/digitallibrary/bookapi/BookApiProperties.scala]: s/Crowidin/Crowdin

File: src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala [hunk 2]
```diff
@@ -44,23 +46,39 @@ trait TranslationService {
         translatedChapters  <- Try(chapterFiles.map(chapter => crowdinClient.fetchTranslatedChapter(chapter, inTranslation.crowdinToLanguage)))
         persisted           <- inTransaction { implicit session =>
           for {
-            persisted       <- Try(writeService.updateTranslation(existingTranslation.copy(title = translatedMetadata.title, about = translatedMetadata.description)))
-            mergedChapters  <- Try(mergeService.mergeChapters(persisted, translatedChapters.filter(_.isSuccess).map(_.get)))
-            _               <- Try(mergedChapters.map(ch => writeService.updateChapter(ch)))
+            existingTranslationWithContrib <- Try(extractContributors(translatedMetadata, existingTranslation))
+            persisted                      <- Try(writeService.updateTranslation(existingTranslationWithContrib.copy(title = translatedMetadata.title, about = translatedMetadata.description)))
+            mergedChapters                 <- Try(mergeService.mergeChapters(persisted, translatedChapters.filter(_.isSuccess).map(_.get)))
```
Comment (runarfu) [inline src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala]: `filter(_.isSuccess).map(_.get)` kan erstattes med `flatten`?

File: src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala [hunk 3]
```diff
@@ -72,20 +90,27 @@ trait TranslationService {
...
-    def updateTranslationStatus(projectIdentifier: String, language: LanguageTag, fileId: String, status: TranslationStatus.Value): Try[InTranslationFile] = {
-      translationDbService.fileForCrowdinProjectWithFileIdAndLanguage(projectIdentifier, fileId, language) match {
-        case None => Failure(new NotFoundException(s"No translation for project $projectIdentifier, language $language and file_id $fileId"))
-        case Some(file) =>
-          translationDbService.updateTranslationStatus(file, status)
+    def extractContributors(bookMetaData: BookMetaData, translation: Translation): Translation = {
+      bookMetaData.translators match {
+        case None => translation
+        case Some(translators) => {
+          val persons = translators.replace(CrowidinTranslatorPlaceHolder, "").split(",").filter(!_.isEmpty).map(_.trim).map(writeService.addPerson)
```
Comment (runarfu) [inline src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala]: `!_.isEmpty === _.nonEmpty`, eventuelt `filterNot(_.isEmpty)`

File: src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala [hunk 4]
```diff
@@ -235,8 +227,9 @@ class TranslationServiceTest extends UnitSuite with TestEnvironment {
     result.failed.get.getMessage contains "No metadata for translation with id" should be (true)
   }
 
-  test("that fetchTranslatedFile returns Failure when translation file is not found") {
+  test("that fetchTranslatedFile returns Failure when no translation files for any languages are not found") {
```
Comment (runarfu) [inline src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala]: "when no translation files are found" or "when translation files are not found"

File: src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala [hunk 5]
```diff
@@ -304,4 +297,55 @@ class TranslationServiceTest extends UnitSuite with TestEnvironment {
...
+    when(chapterRepository.withId(any[Long])(any[DBSession])).thenReturn(Some(TestData.Domain.DefaultChapter))
+    when(mergeService.mergeChapter(any[domain.Chapter], any[TranslatedChapter])).thenReturn(TestData.Domain.DefaultChapter)
+    when(chapterRepository.updateChapter(any[domain.Chapter])).thenReturn(TestData.Domain.DefaultChapter)
+    when(translationDbService.updateTranslationStatus(any[domain.InTranslationFile], any[domain.TranslationStatus.Value])).thenReturn(Success(TestData.Domain.DefaultInTranslationFile))
+
+    val result = service.fetchTranslatedFile("abc", "es", "file-id", domain.TranslationStatus.TRANSLATED)
+    result should be a 'Success
+  }
+
+  test("that exctractContributors adds new contributors and keeps existing contributors to translation") {
```
Comment (runarfu) [inline src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala]: typo extractContributors

File: src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala [hunk 6]
```diff
@@ -304,4 +297,55 @@ class TranslationServiceTest extends UnitSuite with TestEnvironment {
...
+      .thenReturn(person1)
+      .thenReturn(person3)
+
+    when(writeService.addTranslatorToTranslation(eqTo(existingTranslation.id.get), eqTo(person3))).thenReturn(addedContributor)
+
+    val translation = service.extractContributors(bookMetadata, existingTranslation)
+    translation.contributors.size should be (3)
+    translation.contributors.contains(existingContributor1) should be (true)
+    translation.contributors.contains(existingContributor2) should be (true)
+    translation.contributors.contains(addedContributor) should be (true)
```
Comment (runarfu) [inline src/test/scala/io/digitallibrary/bookapi/service/translation/TranslationServiceTest.scala]: This block can be replaced by
```scala
service.extractContributors(bookMetadata, existingTranslation).contributors.toSet should equal(Set(existingContributor1, existingContributor2, addedContributor))
```

**Comment by runarfu** [comment]
Ser bra ut! Bare noen få skrivefeil og et par forslag til kodeforenkling(?)

File: src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala (hunk 2 above)
Comment (kstigen) [reply src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala]: Niks, men med `.flatMap(_.toOption)`

Comment (runarfu) [reply src/main/scala/io/digitallibrary/bookapi/service/translation/TranslationService.scala]: Ah, ok

**Pull request closed without merging**